### PR TITLE
Delta table output connector.

### PR DESCRIPTION
**Pull request by ryzhyk**
A preliminary implementation of an output connector that writes to a delta table.

This commit introduces the notion of an integrated output connector that combines transport and format connectors.  We currently store connector configuration in the `ConnectorConfig::transport` field, with the `format` field now being optional.

This is not yet fully delta lake compatible.  I will keep working on this in the coming weeks, but want to get this in to prevent future conflicts.

Is this a user-visible change (yes/no): no

### CODE REVIEW

File: crates/pipeline-types/src/transport/delta_table.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use crate::config::TransportConfigVariant;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use utoipa::ToSchema;
+
+fn default_max_rows_per_file() -> usize {
+    10_000_000
+}
+
+/// Configuration for a
```
Comment (mihaibudiu) [inline crates/pipeline-types/src/transport/delta_table.rs:10]: unfinished comment

File: crates/adapters/src/integrated/delta_table/output.rs [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+                                                    "error writing parquet file '{batch_uri}': {e}"
+                                                ),
+                                                false,
+                                            )))
+                                            .await;
+                                    }
+                                }
+                            }
+                            Err(e) => {
+                                let _ = response_sender.send(Err((anyhow!("error writing parquet file with {num_rows} rows to '{batch_uri}': {e}"), false))).await;
```
Comment (mihaibudiu) [inline crates/adapters/src/integrated/delta_table/output.rs:231]: it's nice to report errors, but I have no idea what happens to the system when one of these hits
maybe a new file should be started?

File: crates/adapters/src/integrated/delta_table/output.rs [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+            if !cursor.val_valid() {
+                cursor.step_key();
+                continue;
+            }
+            let mut w = cursor.weight();
+            if !(-MAX_DUPLICATES..=MAX_DUPLICATES).contains(&w) {
+                bail!("Unable to output record with very large weight {w}. Consider adjusting your SQL queries to avoid duplicate output records, e.g., using 'SELECT DISTINCT'.");
+            }
+            if w < 0 {
+                // TODO: we don't support deletes in the parquet format yet.
```
Comment (mihaibudiu) [inline crates/adapters/src/integrated/delta_table/output.rs:318]: bleah

File: crates/adapters/src/integrated/delta_table/output.rs (hunk 2 above)
Comment (ryzhyk) [reply crates/adapters/src/integrated/delta_table/output.rs:231]: It does start a new file. We can add retries, but if that fails your only options are to live with incomplete data or rebuild the output from scratch (which we don't currently support).  I'll add a TODO to add logic for retries.

File: crates/adapters/src/integrated/delta_table/output.rs (hunk 3 above)
Comment (ryzhyk) [reply crates/adapters/src/integrated/delta_table/output.rs:318]: This should be coming soon with proper delta table support.

File: crates/pipeline-types/src/transport/delta_table.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+fn default_max_rows_per_file() -> usize {
+    10_000_000
+}
+
+/// Delta table output connector configuration.
+#[derive(Debug, Clone, Eq, PartialEq, Deserialize, Serialize, ToSchema)]
+pub struct DeltaTableWriterConfig {
+    /// Table URI.
+    pub uri: String,
+    /// Storage options for configuring backend object store.
```
Comment (snkas) [inline crates/pipeline-types/src/transport/delta_table.rs:15]: Might there be a reference which options can be set here?

**Review by snkas — COMMENTED** [review]
Looks good overall, might there be a demo script/setup to run locally to try it out end-to-end?

File: crates/dbsp/Cargo.toml [hunk 5]
```diff
@@ -107,7 +107,7 @@ reqwest = { version = "0.11.11", features = ["blocking"] }
 chrono = { git = "https://github.com/chronotope/chrono", rev = "849932", features = ["rkyv-64", "serde"] }
 env_logger = { version = "0.11.2", features = ["color"] }
 serde_json = "1.0.87"
-ijson = { git = "https://github.com/abhizer/ijson.git", branch = "rkyv" }
+ijson = { git = "https://github.com/abhizer/ijson.git", branch = "master" }
```
Comment (gz) [inline crates/dbsp/Cargo.toml:110]: we should put an explicit revision here as any push to this may or may not break the build

**Review by gz — APPROVED** [review]

**Comment by ryzhyk** [comment]
> Looks good overall, might there be a demo script/setup to run locally to try it out end-to-end?

I plan to add that, along with docs for storage options in the next PR.  This is not really a complete or useful connector yet, but I want to merge it to avoid conflicts in the future and also so we can use it in customer demos.

**Pull request merged**